spectrum-offchain-cardano/src/tx_submission.rs

```diff
@@ -140,39 +138,148 @@ where
     }
 }
 
-#[derive(Debug)]
-pub enum TxRejected {
-    MissingInputs(HashSet<OutputRef>),
-    Unknown,
-}
-
-impl TryFrom<TxRejected> for HashSet<OutputRef> {
+impl TryFrom<RejectReasons> for HashSet<OutputRef> {
     type Error = &'static str;
-    fn try_from(value: TxRejected) -> Result<Self, Self::Error> {
-        match value {
-            TxRejected::MissingInputs(inputs) => Ok(inputs),
-            TxRejected::Unknown => Err("Unknown rejection"),
+    fn try_from(value: RejectReasons) -> Result<Self, Self::Error> {
+        let mut outputs = HashSet::new();
+
+        for reason in value.0 {
+            if let ProtocolRejectReason::BadInputsUtxo(inputs) = reason {
+                let o = inputs.into_iter().map(|TxInput { tx_hash, index }| {
+                    let tx_hash = *tx_hash;
+                    OutputRef::new(tx_hash.into(), index)
+                });
+                outputs.extend(o);
+            }
         }
-    }
-}
 
-impl Display for TxRejected {
-    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
-        match &self {
-            TxRejected::MissingInputs(inputs) => write!(f, "TxRejected::MissingInputs({:?})", inputs),
-            TxRejected::Unknown => write!(f, "TxRejected::Unknown"),
+        if !outputs.is_empty() {
+            Ok(outputs)
+        } else {
+            Err("No missing inputs")
         }
     }
 }
 
 #[async_trait::async_trait]
-impl<const ERA: u16, Tx> Network<Tx, TxRejected> for TxSubmissionChannel<ERA, Tx>
+impl<const ERA: u16, Tx> Network<Tx, RejectReasons> for TxSubmissionChannel<ERA, Tx>
 where
     Tx: Send,
 {
-    async fn submit_tx(&mut self, tx: Tx) -> Result<(), TxRejected> {
+    async fn submit_tx(&mut self, tx: Tx) -> Result<(), RejectReasons> {
         let (snd, recv) = oneshot::channel();
         self.0.send(SubmitTx(tx, snd)).await.unwrap();
         recv.await.expect("Channel closed").into()
     }
 }
+
+#[derive(Debug, Clone, derive_more::Display, derive_more::From)]
+#[display(fmt = "RejectReasons: {:?}", "_0")]
+pub struct RejectReasons(pub Vec<ProtocolRejectReason>);
+
+#[derive(Debug, Clone, derive_more::Display)]
+pub enum ProtocolRejectReason {
+    /// Missing/bad TX inputs
+    #[display(fmt = "ProtocolRejectReason::BadInputsUtxo: {:?}", "_0")]
+    BadInputsUtxo(Vec<TxInput>),
+    #[display(
+        fmt = "ProtocolRejectReason::ValueNotConserved: Consumed {:?}, produced {:?}",
+        consumed,
+        produced
+    )]
+    ValueNotConserved { consumed: Value, produced: Value },
+    #[display(fmt = "ProtocolRejectReason::MissingVKeyWitnesses: {:?}", "_0")]
+    MissingVKeyWitnesses(Vec<pallas_crypto::hash::Hash<28>>),
+    #[display(fmt = "ProtocolRejectReason::MissingScriptWitnesses: {:?}", "_0")]
+    MissingScriptWitnesses(Vec<pallas_crypto::hash::Hash<28>>),
+    /// A generic error that cannot be clearly identified.
+    Unknown(String),
```

Comment:
Do we really need this? The only error that can be handled by bot is MissingInputs, everything else should just be logged and ignored.

Reply:
Don't strictly need it, but if you look at `impl From<ShelleyLedgerPredFailure> for ProtocolRejectReason` below, the raw representations from `cardano-ledger` are very nested and make it look more complex than it should be.

Reply:
The rejection structure itself is not used any further from this point, in tlb logic it's abstract and is converted into missing inputs via From<Rejection> for HashSet<Oref>.

Reply:
Ok got it, removed the type.